packages-exp/auth-exp/src/core/user/reload.test.ts

```diff
@@ -0,0 +1,158 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { expect, use } from 'chai';
+import * as chaiAsPromised from 'chai-as-promised';
+import * as sinon from 'sinon';
+import * as sinonChai from 'sinon-chai';
+
+import { mockEndpoint } from '../../../test/api/helper';
+import { testUser } from '../../../test/mock_auth';
+import * as fetch from '../../../test/mock_fetch';
+import { Endpoint } from '../../api';
+import {
+  APIUserInfo,
+  ProviderUserInfo
+} from '../../api/account_management/account';
+import { UserInfo } from '../../model/user';
+import { ProviderId } from '../providers';
+import { _reloadWithoutSaving, reload } from './reload';
+
+use(chaiAsPromised);
+use(sinonChai);
+
+const BASIC_USER_INFO: UserInfo = {
+  providerId: ProviderId.FIREBASE,
+  uid: 'uid',
+  email: 'email',
+  displayName: 'displayName',
+  phoneNumber: 'phoneNumber',
+  photoURL: 'photoURL'
+};
+
+const BASIC_PROVIDER_USER_INFO: ProviderUserInfo = {
+  providerId: ProviderId.FIREBASE,
+  rawId: 'uid',
+  email: 'email',
+  displayName: 'displayName',
+  phoneNumber: 'phoneNumber',
+  photoUrl: 'photoURL'
+};
+
+describe('reload()', () => {
+  beforeEach(fetch.setUp);
+  afterEach(fetch.tearDown);
+
+  it('sets all the new properties', async () => {
+    const serverUser: APIUserInfo = {
+      localId: 'localId',
+      displayName: 'displayName',
+      photoUrl: 'photoURL',
+      email: 'email',
+      emailVerified: true,
+      phoneNumber: 'phoneNumber',
+      tenantId: 'tenantId',
+      createdAt: 123,
+      lastLoginAt: 456
+    };
+
+    mockEndpoint(Endpoint.GET_ACCOUNT_INFO, {
+      users: [serverUser]
+    });
+
+    const user = testUser('abc', '', true);
+    await _reloadWithoutSaving(user);
+    expect(user.uid).to.eq('localId');
+    expect(user.displayName).to.eq('displayName');
+    expect(user.photoURL).to.eq('photoURL');
+    expect(user.email).to.eq('email');
+    expect(user.emailVerified).to.be.true;
+    expect(user.phoneNumber).to.eq('phoneNumber');
+    expect(user.tenantId).to.eq('tenantId');
+    expect(user.metadata).to.eql({
+      creationTime: '123',
+      lastSignInTime: '456'
+    });
+  });
+
+  it('adds missing provider data', async () => {
+    const user = testUser('abc', '', true);
+    user.providerData = [{ ...BASIC_USER_INFO }];
```

Comment:
isn't this just [BASIC_USER_INFO]?

Reply:
No, given that things in the user object do get mutated, we want to make sure we're using a copy. The objects in providerData _shouldn't_ be mutated, but then again that's why we have tests :)